AMM/SPECIFICATION.md

```diff
@@ -0,0 +1,180 @@
+
+If you are interested in a functional overview then this is the section for you.
+
+## Actions that users are able to perform
+
+This sub-section details what a user is able to do e.g. click a button and "x, y, z" happens.
+
+### AMM Core Functionality
+
+#### `initialize()`
+
+1. Allows specifying the legitimate exchange contract implementation that the AMM will operate with Requires
+    1. The bytecode root of the desired exchange contract implementation
+
+#### `add_pool()`
+
+1. Allows an asset pair to be registered to the AMM 
+    1. If the AMM is initialized Requires
+    2. Requires
+        1. The identifiers of the two assets
+        2. The exchange contract identifier that is also the identifier of the liquidity pool asset for the given pair if the exchange contract
+            1. Is legitimate, i.e., the bytecode root matches
+            2. Defines the pool for the provided asset pair
+
+### AMM State Checks
+
+#### `pool()`
+
+1. Returns the registered exchange contract identifier for an asset pair Requires
+    1. The identifiers of the two assets
+
+### Exchange Core Functionality
+
+#### `constructor()`
+
+1. Allows specifying the asset pair for which the pool will be managed through the exchange contract 
+    1. If the asset pair for the exchange contract has not already been set 
+    2. Requires two different asset identifiers
+
+#### `deposit()`
+
+1. Allows depositing an asset to the contract (to later be used for adding liquidity or be withdrawn) 
+    1. If the asset pair of the pool is set 
+    2. Requires any amount of either asset in the pair
+
+#### `add_liquidity()`
+
+1. Allows minting liquidity pool assets by using up the deposits when possible 
+    1. If
+        1. The asset pair of the pool is set
+        2. The deadline has not passed
+        3. Deposit amounts of both assets are sufficient
+        4. The resulting liquidity amount is more than the minimum liquidity for a pool to exist. 
+            > **NOTE** This lower limit for liquidity can be modified as desired per asset pair and exists to protect against excessive slippage.
+    2. Requires 
+        1. The desired liquidity amount
+        2. A deadline, i.e., block height limit
+
+#### `remove_liquidity()`
+
+1. Allows burning liquidity pool assets Requires 
+    1. If
+        1. The asset pair of the pool is set
+        2. The deadline has not passed
+        3. The amount of liquidity pool assets are sufficient
+    2. Requires
+        1. The minimum amounts of both assets to receive after burning
+        2. A deadline, i.e., block height limit
+
+#### `withdraw()`
+
+1. Allows withdrawing previously deposited assets without using them to add liquidity 
+    1. If
+        1. The asset pair of the pool is set
+        2. The deposited amount of the asset is sufficient
+    2. Requires an amount of either asset to withdraw
+
+#### `swap_with_exact_input()`
+
+1. Allows selling an exact amount of an asset for the other asset 
```

Comment:
Without looking at the code it's unclear which / what asset has some exact amount and what this `other` asset is. Can you be more specific i.e. if there are two assets that are provided as arguments then state the amount of asset A and that it gets swapped for the specified asset B etc.

Reply:
The input and output assets can be either one of asset A or asset B. This is specified with the function call.